doc: documenting the boot protocol

dpsmith opened the pull request:
This is an attempt to formalize the boot protocol while also proposing
a new name for this implementation of the Secure Loader for SKINIT.

Signed-off-by: Daniel P. Smith <dpsmith@apertussolutions.com>

dpsmith committed:
doc: documenting the boot protocol

This is an attempt to formalize the boot protocol while also proposing
a new name for this implementation of the Secure Loader for SKINIT.

Signed-off-by: Daniel P. Smith <dpsmith@apertussolutions.com>

andyhhp commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+Security Kernel Loader
+######################
+
+The Security Kernel Loader (SKL) is an implementation of a Secure Loader for
+the SKINIT CPU instruction. The SKL loader takes a minimal approach relying on
```
I'd phrase this as "The SKL is an implementation of AMD's SLB spec for the SKINIT CPU instruction and DRTM measurements" or similar.

The fact this is an implementation of AMD's spec really needs to be in the first sentence, because it is the single most important piece of information people reading this doc need to know.

dpsmith replied:
I was being gun-shy about saying AMD here since in theory this should work with Hygon. But will gladly reword it.

rossphilipson replied:
Is Hygon basically AMD? I ask in ignorance; from what I can see, it is an x86 partnership between AMD and China

andyhhp commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+necessary size for the allocation will be found in the SKL header. The DRTM
+Preamble will place the SKL at the beginning of the allocation.
+
+Below is the layout of how the memory allocation is segmented::
+
+      +---     Security Kernel Loader    ---+
+      |                                     |
+      v                                     v
+      +--------+----------+-------------+-------------+--------------+
+      | Header |   Code   |  Boot Tags  | IOMMU Table |  Log Buffer  |
+      +--------+----------+-------------+-------------+--------------+
+      ^                                 ^                            ^
+      |                                 |                            |
+      +                                 +                            +
+  sklbuf_base                          60kb                   header.skl_size
```
"how the memory is laid out", but do we really want to include this?  It is liable to get stale quite quickly.

dpsmith replied:
well, hopefully it won't go stale quickly. This is an attempt to get to a stable ABI.

krystian-hebel replied:
I agree with @andyhhp, e.g. we were discussing generating page tables in the code, in which case we could put them just below IOMMU table, which should help with fragmentation caused by alignment. If we want to have a layout, I think it would be better to stick to code/data/bss regions (maybe also boot tags?) and to not go into much details. Bootloader doesn't really have to know how SKL uses that memory, it only has to know how much it has to reserve.

`header.skl_size` may be misleading because it is different than range marked above, although I'm not sure which one of those to change. It may be worth mentioning that the end of `Security Kernel Loader` is always at 64K and this is hardware limitation, not something that we designed. Another critical piece of information that is missing: `sklbuf_base` must be aligned to 64K.

dpsmith replied:
Ah,I see the confusion. The IOMMU Table region I am proposing is not a static/compile time generated region. This layout was specified specifically with the assumption that the IOMMU tables would be generated at run time. During runtime the SKL would locate the end of the boot tags and then take the first 4k aligned address after, provided it was not outside the 64k SKINIT protected region. Thus why i was trying to put the restriction on the boot protocol that boot tags could not exceed past the 60k offset. What needs to be determined at compile time is how much space in total does the SKL need for the SLB (64k) + additional external space, and communicate that to the bootloader via the header. Now we could rename header.skl_size to header.alloc_req or something similar but what is important is that this protocol explains how the boot loader is to use the information. As for the diagram, I think it should stay but with detail removed to reduce confusion and not mix internal details with information needed by external entities.

andyhhp commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+::
+
+    typedef struct __packed sl_header {
+        u16 entry_point;        /* Code Entry */
+        u16 boot_tags_offset;   /* RoT Measurement Length */
```
These should be the SLB names, and in particular the second one is the measurement length, not the boot tags offset.

We do not want to tie ourselves to an ABI where the boot tags immediately follow the end of measurement.  It happens to be how we lay things out now, but we really ought to have a 3rd pointer to the tags so we can alter the layout if needs be.

dpsmith replied:
I was split on this one. Since it was currently set to boot params, I left it for the time being to get comments. I do like the suggestion of adding a field to point to the boot tags.

krystian-hebel replied:
I was going back and forth between having a separate field for boot tags offset and not having it. In the end decided against it because at that point one additional `u16` was just enough to increase the size another 4K because of page_data alignment :slightly_smiling_face:  I agree on changing the name to SLB one, and probably on adding a separate field too. We can put stack between measured end of part and beginning of tags, that way it would have known size and wouldn't be unnecessarily measured.

dpsmith replied:
I like the idea for moving the stack, which also goes back to @andyhhp point that for external entities, i.e. bootloader, they do not need to know those details, they need to know how much memory to reserve, and where the boot tags start so they can fill in the appropriate information.

andyhhp commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+two additional fields which can be seen in the following representative struct,
+::
+
+    typedef struct __packed sl_header {
```
Technically speaking, `__packed` is something we should not have here.  It doesn't affect the layout of the struct, other than reducing the alignment to 1 which prohibits you from writing code like `u32 *entry_point_ptr = &sl_header.entry_point;`

dpsmith replied:
Ack, can drop here and maybe even to a PR to drop it from the current code.

andyhhp commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+SKL Allocation Size
+-------------------
+
+The length of Boot Tags are a runtime-determined size but cannot exceed the
+last possible start address for the IOMMU Table, that being 61440 byte offset
+from the beginning of the allocation. For build time calculation of the SKL
+allocation size, the IOMMU table will be considered to start at offset 61440.
+This renders the build time calculation to,
```
The IOMMU isn't an implementation detail, and shouldn't be here.

The length of the boot tags is derived from the requirement to be outside of the measured region, but within the 64k DMA-protected region, and not overlapping with any other bits of that space that SKL is using.

dpsmith replied:
I get where you are going but this document is to tell boot loaders how then need to setup the loader and pass the necessary information. Since at least the first page of IOMMU and all of the boot params needs to be in the 64k page, they need to understand that when filling in the boot tags, they cannot exceed the last page.

krystian-hebel replied:
Is there anything that bootloader can do if there is not enough space for tags? If not, then there is no need to go into details. I think a build-time test in SKL for a minimal amount of space required for tags should suffice. We also may add another field to the header, but we definitely shouldn't hardcode it on the bootloader's side, since we may put other page-aligned data there, as mentioned in one of my previous comments.

andyhhp commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+identifier and the version for the SKL.
+::
+
+    typedef struct __packed skl_info {
+        u8  uuid[16]; /* 78 f1 26 8e 04 92 11 e9  83 2a c8 5b 76 c4 cc 02 */
+        u32 version;
+    } skl_info_t;
```
This is version 1.  Future versions may extend this structure in a backwards compatible manor.

dpsmith replied:
yah, should have a statement such as that.

andyhhp commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+        u8 digest[];
+    } __packed;
+
+IOMMU Table
+===========
+
+The IOMMU Table will be constructed at the first 4k page-aligned address after
+the Boot Tags but no further than 61440 bytes into the SKL. 
```
All implementation details, and not applicable here.

What may be relevant is that handover from SKL to the next step in the boot chain shall be performed with the IOMMU(s) enabled and blocking everything, and that the next step in the boot chain must take over control of the IOMMU.

dpsmith replied:
On this one I would have to agree, not relevant to the boot protocol but to the implementation.

andyhhp commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+measurement value's source, this is important as it will change how the
+attester must replay the log to match a TPM quote.
+
+Kernel Log
+----------
+
+The kernel log is an optional log region for the SLK to store information and
+debugging messages. This log will be located after the Event Log and will be
+given the remainder of the Log Buffer region.
```
Surely we want to have explicit base (or offset from the start of SKL) and length fields for the logs?

dpsmith replied:
Yah, it will need be otherwise it would require external entities to know/understand the memory layout.

krystian-hebel commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+Memory Layout
+*************
+
+The DRTM Preamble is responsible for allocating a contiguous block of physical
```
Does it actually have to be continuous? IOMMU has to be, but log can be put anywhere (probably best to limit it to <4GB).

dpsmith replied:
It makes things simpler otherwise we would have to expand the header each time we add another section that the boot loader could allocate independently. This way, the bootloader just needs to find a single chunk in memory layout table and then the SKL can dice it up however it needs.

krystian-hebel commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+allocation size, the IOMMU table will be considered to start at offset 61440.
+This renders the build time calculation to,
+
+    skl_size = 61440 + PAGETABLE_SIZE + LOGBUFFER_SIZE
+
+Where the default for PAGETABLE_SIZE is 28kb and LOGBUFFER_SIZE is 4kb.
```
Which pagetable is it, CPU one? If yes, then this is wrong, we can't put it outside of SLB.

krystian-hebel commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+        u8 digest[];
+    } __packed;
+
+Kernel Log Tag
+^^^^^^^^^^^^^^
+
+The Kernel Log tag provides a means to control/configure logging done by SKL
+itself. Specifically to enable/disable and set priority of the logging to be
+done by SKL.
+::
+
+    struct skl_tag_crtm_hash {
+        struct skl_tag_hdr hdr;
+        u8 enable;
+        u8 priority;
+    } __packed;
```
You already mentioned it somewhere, but the location of log buffer is missing - just a reminder.

krystian-hebel replied:
Also, copy-paste of struct name from the previous tag.

krystian-hebel commented on doc/boot_protocol.rst:
```diff
@@ -0,0 +1,274 @@
+Log Buffer
+==========
+
+The Log Buffer region contains the TPM Event Log and optionally a kernel log
```
Currently for TPM event log we search for DRTM ACPI table that should have pointer and size of event log, and if not found, GRUB allocates memory for it and passes the tag, so it should be always available. Why do we want to change this logic?

dpsmith replied:
My mistake, we do want to keep that.

dpsmith commented:
I have trimmed out the implementation details, addressed the comments, and made a few small expansions. I have pushed the changes to the boot protocol doc and have the implementation details in an in progress document. Even if everyone agrees with this revision, I would like to hold this PR until I can get the implementation document done so that we can ensure they are aligned with each other before merging.
